Feat/add discovery client

aschmahmann opened the pull request:
Added a stateful client that implements the libp2p discovery interface.

aschmahmann committed:
Switched from gx to go mod and started using go-libp2p-core interfaces

aschmahmann committed:
Add stateful discovery client

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,134 @@
+
+	// Default is minimum duration
+	if ttl == 0 {
+		ttlSeconds = 120
```
let's make that 2 hours, 2 minutes is too short to be useful

aschmahmann replied:
Change the default or the minimum?

vyzo replied:
default should not be the minimum duration, but a reasonable default; hence two hours.

aschmahmann replied:
Done

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,134 @@
+	if ttl == 0 {
+		ttlSeconds = 120
+	} else {
+		ttlSeconds = int(math.Round(ttl.Seconds()))
```
why are you rounding an integer?

aschmahmann replied:
Duration.Seconds() returns a float (e.g. 1.5 seconds instead of 1 or 2). I decided rounding was our best bet, but we could do Floor instead if that's better.

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,134 @@
+		ttlSeconds = int(math.Round(ttl.Seconds()))
+	}
+
+	if err := c.rp.Register(ctx, ns, ttlSeconds); err != nil {
```
I think we have changed the protocol to return the actual TTL, let's return it from the  `RendezvousPoint` interface and use that.

aschmahmann replied:
I'm not sure I understand what you want to be done here. The `Register` functions in `RendezvousPoint` do not return TTL, I assume because the one implementation of `RendezvousPoint` doesn't modify the input TTL at all.

Would you like me to modify the `RendezvousPoint` and `RendezvousClient` interfaces to return the TTL? If so would you like it as an `int`, or `time.Duration`?

vyzo replied:
Well, the returned TTL was added _after_ the rendezvous point interface was written.
Let's change the interface to return it.

aschmahmann replied:
Changed the interface, I didn't get your feedback on `int` vs `time.Duration` so I went with `time.Duration` to be more compatible with the discovery interface

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,134 @@
+	// Discover new records if we don't have enough
+	var discoveryErr error
+	if newCacheSize < limit {
+		if discoveryRecords, _, err := c.rp.Discover(ctx, ns, limit, nil); err == nil {
```
We should probably loop doing `rp.Discover` by using the cookie so that we get all peers.
This will stop just at the first set of peers and never discover any more.

aschmahmann replied:
We talked about the max limit on this type of blocking client being about 1000 records which conveniently is the maximum limit for the rendezvous server. So one page should be enough until we do a non-blocking client.

vyzo replied:
Let's still use the cookie to always make progress even if we only do a query at a time.

aschmahmann replied:
To clarify, because of the internal caching utilizing the cookie will only help us make progress if we query beyond `limit - newCacheSize`. However, we can definitely use the cookie to help in this case. It should also serve to limit the data returned from repeated `limit=MaxLimit` queries.

aschmahmann replied:
This should be implemented now. However, the way I randomly select items to return is not nearly as efficient as it could be. Given that this is just on the client side and the list sizes are small (<1000) it my not be necessary to get it fixed up now, however any suggestions are welcome.

vyzo reviewed (changes requested):
It's a great start.  
The `FindPeers` method will need to be modified however so that it stores the cookie and iterates, otherwise it will always get just the first set of peers.

aschmahmann committed:
RendezvousPoint and RendezvousClient now return the server's TTL on Registered records.
Default TTL for discovery client increased
discovery client now utilizes server cookie for added efficiency

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,148 @@
+	mux        sync.Mutex
+}
+
+func NewRendezvousDiscoveryClient(host host.Host, rendezvousPeer peer.ID) discovery.Discovery {
```
This is a mouthful, let's call it just `NewRendezvousDiscovery`

aschmahmann replied:
Sure

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,148 @@
+	"time"
+)
+
+type rendezvousDiscoveryClient struct {
```
let's drop the client from the name.

aschmahmann replied:
Sure

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,148 @@
+	cache.mux.Unlock()
+
+	// Discover new records if we don't have enough
+	var discoveryErr error
```
just `err`.

aschmahmann replied:
If I name this `err` I need to rename the internal variable currently named `err` or we run into scoping problems. If you'd like to name both that's fine with me.

vyzo replied:
why is this a different error?

aschmahmann replied:
it's not a different error it's just an interaction between scoping and the way `=` and  `:=` work with multiple return values. I can just invert this and name the inner one `discoveryErr`, `terr`, or whatever and then replace `discoveryErr=err` with `err=discoveryErr` no var declaration required. This sufficient?

vyzo replied:
you can declare a variable first before assigning, this will solve your scoping problem.

aschmahmann replied:
Ok, I'll need two preassignment declarations (one for `records` and one for `newCookie`) but that's fine with me.

vyzo replied:
struct copy is cheap, don't worry about it.

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,148 @@
+	}
+}
+
+func (c *rendezvousDiscoveryClient) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
```
you are taking and releasing the lock all the time here, better to just do it once and have the whole method under lock.

aschmahmann replied:
I guess that's fine. It's super unfortunate that we have an interface that returns channel that is synchronous (sort of an implicit contract break).

Locking the cache during `Discover` may end up increasing latency to return. However, it saves us from a few annoying corner cases, so fine by me.

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,148 @@
+			}
+			for i := range discoveryRecords {
+				rec := &discoveryRecords[i]
+				rec.Ttl += currentTime
```
hrm, that's no longer a ttl but rather an expiry.

aschmahmann replied:
Sure, although I've seen the terminologies used interchangeably. We really want an expiry here though. Would you prefer me to create a new private struct for use internally in `RendezvousDiscovery`?

vyzo replied:
yeah, make a struct.

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,148 @@
+			if cache.cachedRegs == nil {
+				cache.cachedRegs = make(map[peer.ID]*Registration)
+			}
+			for i := range discoveryRecords {
```
`for i, rec := range direcoveryRecords { ... }`

aschmahmann replied:
That change actually breaks the below code unless you have a new variable for the modified rec because `Discover` returns `[]Registration` and not `[]*Registration`.

Modifying `rec` creates a new `Registration` that needs a variable anyway, so why not just call `rec` the pointer to the `Registration` we need?

vyzo replied:
maybe we need a better algorithm, but ok.

aschmahmann replied:
I think this is generally a problem with arrays of structs (instead of pointers to structs). Given that you'd like a new expiry struct anyway I can now iterate over `i, rec` however I'm pretty sure it'll result in an extra copy of the fields in `rec`. Any preferences on if I should change it to `i, rec`?

vyzo replied:
you can't mutate this way.

aschmahmann replied:
yes, that was the problem when we were reusing the struct. Since we will be using a new struct we're copying the data anyway instead of mutating

vyzo replied:
struct copy is cheap, don't worry about it. So yeah, do it with a proper range loop.

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,148 @@
+	// Discover new records if we don't have enough
+	var discoveryErr error
+	if newCacheSize < limit {
+		if discoveryRecords, newCookie, err := c.rp.Discover(ctx, ns, limit, cookie); err == nil {
```
call `discoveryRecords` just `records`.

aschmahmann replied:
Sure

vyzo commented on discovery_client.go:
```diff
@@ -0,0 +1,148 @@
+
+	// Randomize and fill channel with available records
+	cache.mux.Lock()
+	sendQuantity := len(cache.cachedRegs)
```
what is `sendQuantity`???? That's a horrible name for a variable :p

aschmahmann replied:
the quantity of AddrInfos that we're sending

aschmahmann replied:
If you have a better name feel free to suggest.

aschmahmann replied:
@vyzo suggested to use `count`, which is fine by me.

vyzo reviewed (commented):
This still needs work...
Also, can you call the discovery module file just `discovery`? `discovery_client` is a mouthful.
And try to shorten variable/type names as well.

aschmahmann commented:
> This still needs work...
> Also, can you call the discovery module file just `discovery`? `discovery_client` is a mouthful.
> And try to shorten variable/type names as well.

No problem, I've found that longer names are more self descriptive (and not out of style in comparison to a number of Golang packages) and that with autocomplete they aren't actually a pain. However, I also have no interest in debating variable names as long as they are descriptive enough to get the point across so that interested reviewers can propose new names for them it's a simple approve from me.

aschmahmann committed:
fixed compile error from previous commit + code refactoring

vyzo commented on discovery.go:
```diff
@@ -0,0 +1,147 @@
+	rngMux    sync.Mutex
+}
+
+type discoveredPeerCache struct {
```
call this `discoveyCache`

vyzo commented on discovery.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+type discoveredPeerCache struct {
+	cachedRecs map[peer.ID]*record
```
and this just `recs`, it's a cache and these are named records.

vyzo commented on discovery.go:
```diff
@@ -0,0 +1,147 @@
+		ttlSeconds = int(math.Round(ttl.Seconds()))
+	}
+
+	if returnedTTL, err := c.rp.Register(ctx, ns, ttlSeconds); err != nil {
```
`rttl` instead of `returnedTTL`, or just reuse the `ttl` variable.

vyzo reviewed (commented):
this is looking good.

aschmahmann committed:
replaced sync.Map with map + RW mutex. small refactors

aschmahmann commented:
@vyzo switched sync.Map with Map + RWMutex like we did in pubsub
